apiserver/facades/agent/caasoperator/watcher.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caasoperator
+
+import (
+	"github.com/juju/errors"
+	"gopkg.in/juju/worker.v1"
+	"gopkg.in/juju/worker.v1/catacomb"
+
+	corewatcher "github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/state"
+)
+
+type unitIDWatcher struct {
+	catacomb catacomb.Catacomb
+	out      chan []string
+	src      corewatcher.StringsWatcher
+	model    watcherModelInterface
+}
+
+type watcherModelInterface interface {
+	Containers(providerIds ...string) ([]state.CloudContainer, error)
+}
+
+func newUnitIDWatcher(model watcherModelInterface, src corewatcher.StringsWatcher) (*unitIDWatcher, error) {
+	w := &unitIDWatcher{
+		out:   make(chan []string),
+		src:   src,
+		model: model,
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{src},
+	})
+	return w, err
+}
+
+func (w *unitIDWatcher) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+func (w *unitIDWatcher) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *unitIDWatcher) Stop() error {
+	w.Kill()
+	return w.Wait()
+}
+
+func (w *unitIDWatcher) Changes() <-chan []string {
+	return w.out
+}
+
+func (w *unitIDWatcher) Err() error {
+	return w.catacomb.Err()
+}
+
+func (w *unitIDWatcher) loop() error {
+	defer close(w.out)
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case providerIDs, ok := <-w.src.Changes():
+			if !ok {
+				return errors.Errorf("event watcher closed")
+			}
+			if len(providerIDs) == 0 {
+				w.out <- nil
+				continue
+			}
+			containers, err := w.model.Containers(unique(providerIDs)...)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			providerIDToUnitName := map[string]string{}
+			for _, containerInfo := range containers {
+				providerIDToUnitName[containerInfo.ProviderId()] = containerInfo.Unit()
+			}
+			var result []string
+			for _, providerID := range providerIDs {
+				if unitName, ok := providerIDToUnitName[providerID]; ok {
+					result = append(result, unitName)
+				}
+			}
+			w.out <- result
+		}
+	}
+}
+
+func unique(in []string) []string {
```

Comment:
We have a set.NewStrings() in juju/collections/set that can replace this to dedup the ids